packages/jupyter-ai/jupyter_ai/chat_handlers/tools.py

```diff
@@ -0,0 +1,260 @@
+            list: A list of function names.
+        """
+        try:
+            with open(tools_file_path) as file:
+                content = file.read()
+                tree = ast.parse(content)
+                tools = []
+                for node in ast.walk(tree):
+                    if isinstance(node, ast.FunctionDef):
+                        for decorator in node.decorator_list:
+                            if (
+                                isinstance(decorator, ast.Name)
+                                and decorator.id == "tool"
+                            ):
+                                tools.append(node.name)
+            return tools  # this is a list
+        except FileNotFoundError as e:  # to do
+            self.reply(f"Tools file not found at {tools_file_path}.")
+
+    def toolChat(self, query):
+        print("TOOL CHAT", query)
+        for chunk in self.app.stream(
+            {"messages": [("human", query)]}, stream_mode="values"
+        ):
+            response = chunk["messages"][-1].pretty_print()
+        return response
+
+    ##### MAIN FUNCTION #####
+    def useLLMwithTools(self, query):
+        """
+        LangGraph documentation : https://langchain-ai.github.io/langgraph/tutorials/introduction/
+        The code below:
+        1. Extracts the function names in the custom tools file
+        2. Adds the tools to the Tool Node
+        3. Binds the Tool Node to the LLM
+        4. Sets up a basic LangGraph with nodes and edges
+        5. Compiles the graph into a runnable app
+        6. This function is then called with a prompt
+        Every time a query is submitted the langgraph is rebuilt in case the tools file has been changed.
+        """
+
+        # Calls the requisite tool in the LangGraph
+        def call_tool(state: MessagesState):
+            messages = state["messages"]
+            response = self.model_with_tools.invoke(messages)
+            return {"messages": [response]}
+
+        # Get all tool objects from the tool files
+        def getTools(file_paths):
+            if len(file_paths) > 0:
+                tool_names = []
+                for file_path in file_paths:
+                    with open(file_path) as file:
+                        exec(file.read())
+                    tool_names = tool_names + self.getToolNames(file_path)
+                tools = [eval(j) for j in tool_names]
+                return tools
+
+        # Get tool file(s), then tools within tool files, and create tool node from tools
+        file_paths = self.getToolFiles(self.tools_file_path)
+        tools = getTools(file_paths)
+        tool_node = ToolNode(tools)
+
+        # Bind tools to LLM
+        # Check if the LLM class takes tools else advise user accordingly.
+        # Can be extended to include temperature parameter
+        try:
+            self.model_with_tools = self.llm.__class__(
+                model_id=self.llm.model_id
+            ).bind_tools(tools)
```

Comment:
This only passes the `model_id` to the constructor, while not passing the other parameters that may be required, e.g. the API keys. The code from `self.create_llm_chain()` should be used:


```py
def create_llm_chain(
    self, provider: Type[BaseProvider], provider_params: Dict[str, str]
):
    unified_parameters = {
        "verbose": True,
        **provider_params,
        **(self.get_model_parameters(provider, provider_params)),
    }
    llm = provider(**unified_parameters)
    ...
```

This section of the code can be split into a private method such that it can be used both here and in `create_llm_chain()`, which minimizes code duplication.

Reply:
This section of the code has been placed in a separate function: 
```
def setup_llm(
        self, provider: Type[BaseProvider], provider_params: Dict[str, str]
    ):
        """Sets up the LLM before creating the LLM Chain"""
        unified_parameters = {
            "verbose": True,
            **provider_params,
            **(self.get_model_parameters(provider, provider_params)),
        }
        llm = provider(**unified_parameters)
        self.llm = llm
        return llm
```
This function is now called from 
```
def create_llm_chain(
        self, provider: Type[BaseProvider], provider_params: Dict[str, str]
    ):
        """Uses the LLM set up to create the LLM Chain"""
        llm = self.setup_llm(provider, provider_params)
        prompt_template = llm.get_chat_prompt_template()
        self.llm = llm

        runnable = prompt_template | llm  # type:ignore
        if not llm.manages_history:
            runnable = RunnableWithMessageHistory(
                runnable=runnable,  #  type:ignore[arg-type]
                get_session_history=self.get_llm_chat_memory,
                input_messages_key="input",
                history_messages_key="history",
                history_factory_config=[
                    ConfigurableFieldSpec(
                        id="last_human_msg",
                        annotation=HumanChatMessage,
                    ),
                ],
            )
        self.llm_chain = runnable
```
using the new `runnable` approach with LCEL.